### ISSUE TITLE
Importing requests very slow on Raspberry Pi

### ISSUE THREAD

**Issue by dagwieers** [Freso/script.module.requests#25]
We found that importing the **requests** library on Raspberry Pi v3 (LibreELEC v9.0.1) is very slow and takes 4.2 seconds. Given that an addon that imports requests is called for every menu, it takes at least 4.2 seconds to navigate in addons that use requests.

All addons are affected (for the menus where requests is imported).

This is easy to reproduce:
```bash
$ export PYTHONPATH=/storage/.kodi/addons/script.module.chardet/lib:/storage/.kodi/addons/script.module.urllib3/lib:/storage/.kodi/addons/script.module.certifi/
lib:/storage/.kodi/addons/script.module.idna/lib:/storage/.kodi/addons/script.module.requests/lib:/storage/.kodi/addons/script.module.beautifulsoup4/lib:/storage/.kodi/a
ddons/script.module.dateutil/lib
$ time python -c 'import requests'
real	0m 4.27s
user	0m 4.08s
sys	0m 0.14s
$ time python -c 'import bs4'
real	0m 0.25s
user	0m 0.21s
sys	0m 0.03s
$ time python -c 'import dateutil'
real	0m 0.12s
user	0m 0.07s
sys	0m 0.04s
```

Other platforms are not affected, both Windows and Android work fine. So it appears related to Raspberry Pi specifically.

Ways to mitigate:
- Move the imports into functions, rather than modules (so it doesn't happen for all menus)
- Move to urllib2 (Python 2) and urllib (Python 3)

Both are quite messy.

**Comment by dagwieers** [comment]
More information on what we tested: https://github.com/pietje666/plugin.video.vrt.nu/issues/169

**Comment by dagwieers** [comment]
Here is some output from profiling the **requests** module:

**test-requests.py**
```python
#!/usr/bin/python

import sys

sys.path.append('/storage/.kodi/addons/script.module.chardet/lib')
sys.path.append('/storage/.kodi/addons/script.module.urllib3/lib')
sys.path.append('/storage/.kodi/addons/script.module.certifi/lib')
sys.path.append('/storage/.kodi/addons/script.module.idna/lib')
sys.path.append('/storage/.kodi/addons/script.module.requests/lib')

import cProfile

pr = cProfile.Profile()
pr.enable()
import requests

requests.get('https://vrtnu-api.vrt.be/suggest?facets[transcodingStatus]=AVAILABLE').json()
pr.disable()
pr.dump_stats('/tmp/profiling_stats')

import pstats
p = pstats.Stats('/tmp/profiling_stats')
p.sort_stats('name')
p.print_stats()

p.sort_stats('cumulative').print_stats(20)
```

This outputs:
```
<snip>
Fri May  3 14:21:15 2019    /tmp/profiling_stats

         224287 function calls (214550 primitive calls) in 5.064 seconds

   Ordered by: cumulative time
   List reduced from 1220 to 20 due to restriction <20>

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        1    0.006    0.006    4.304    4.304 /storage/.kodi/addons/script.module.requests/lib/requests/__init__.py:41(<module>)
        1    0.003    0.003    4.079    4.079 /storage/.kodi/addons/script.module.urllib3/lib/urllib3/__init__.py:3(<module>)
        1    0.007    0.007    4.067    4.067 /storage/.kodi/addons/script.module.urllib3/lib/urllib3/connectionpool.py:1(<module>)
       80    0.001    0.000    3.983    0.050 /usr/lib/python2.7/re.py:192(compile)
       80    0.009    0.000    3.983    0.050 /usr/lib/python2.7/re.py:230(_compile)
       75    0.002    0.000    3.973    0.053 /usr/lib/python2.7/sre_compile.py:567(compile)
        1    0.000    0.000    3.782    3.782 /storage/.kodi/addons/script.module.urllib3/lib/urllib3/packages/rfc3986/__init__.py:23(<module>)
        1    0.002    0.002    3.782    3.782 /storage/.kodi/addons/script.module.urllib3/lib/urllib3/packages/rfc3986/api.py:20(<module>)
        1    0.003    0.003    3.773    3.773 /storage/.kodi/addons/script.module.urllib3/lib/urllib3/packages/rfc3986/iri.py:1(<module>)
        1    0.001    0.001    3.742    3.742 /storage/.kodi/addons/script.module.urllib3/lib/urllib3/packages/rfc3986/misc.py:20(<module>)
       75    0.001    0.000    3.553    0.047 /usr/lib/python2.7/sre_compile.py:552(_code)
  2019/75    0.084    0.000    3.503    0.047 /usr/lib/python2.7/sre_compile.py:64(_compile)
      721    0.086    0.000    3.372    0.005 /usr/lib/python2.7/sre_compile.py:228(_compile_charset)
      721    2.013    0.003    3.284    0.005 /usr/lib/python2.7/sre_compile.py:256(_optimize_charset)
     2572    1.235    0.000    1.235    0.000 {range}
        1    0.000    0.000    0.515    0.515 /storage/.kodi/addons/script.module.requests/lib/requests/api.py:63(get)
        1    0.000    0.000    0.515    0.515 /storage/.kodi/addons/script.module.requests/lib/requests/api.py:16(request)
        1    0.000    0.000    0.506    0.506 /storage/.kodi/addons/script.module.requests/lib/requests/sessions.py:466(request)
        1    0.000    0.000    0.496    0.496 /storage/.kodi/addons/script.module.requests/lib/requests/sessions.py:617(send)
        1    0.000    0.000    0.451    0.451 /storage/.kodi/addons/script.module.requests/lib/requests/adapters.py:394(send)
```

**Comment by dagwieers** [comment]
System call tracing shows the time is consumed after importing *urllib3/packages/rfc3986/abnf_regexp.py*
```
14:30:43.875206 stat64("/storage/.kodi/addons/script.module.urllib3/lib/urllib3/packages/rfc3986/re", 0x7ebc7578) = -1 ENOENT (No such file or directory)
14:30:43.875413 openat(AT_FDCWD, "/storage/.kodi/addons/script.module.urllib3/lib/urllib3/packages/rfc3986/re.so", O_RDONLY|O_LARGEFILE) = -1 ENOENT (No such file or directory)
14:30:43.875639 openat(AT_FDCWD, "/storage/.kodi/addons/script.module.urllib3/lib/urllib3/packages/rfc3986/remodule.so", O_RDONLY|O_LARGEFILE) = -1 ENOENT (No such file or directory)
14:30:43.875816 openat(AT_FDCWD, "/storage/.kodi/addons/script.module.urllib3/lib/urllib3/packages/rfc3986/re.py", O_RDONLY|O_LARGEFILE) = -1 ENOENT (No such file or directory)
14:30:43.876014 openat(AT_FDCWD, "/storage/.kodi/addons/script.module.urllib3/lib/urllib3/packages/rfc3986/re.pyo", O_RDONLY|O_LARGEFILE) = -1 ENOENT (No such file or directory)
14:30:43.876321 stat64("/storage/.kodi/addons/script.module.urllib3/lib/urllib3/packages/rfc3986/abnf_regexp", 0x7ebc7568) = -1 ENOENT (No such file or directory)
14:30:43.876477 openat(AT_FDCWD, "/storage/.kodi/addons/script.module.urllib3/lib/urllib3/packages/rfc3986/abnf_regexp.so", O_RDONLY|O_LARGEFILE) = -1 ENOENT (No such file or directory)
14:30:43.876681 openat(AT_FDCWD, "/storage/.kodi/addons/script.module.urllib3/lib/urllib3/packages/rfc3986/abnf_regexpmodule.so", O_RDONLY|O_LARGEFILE) = -1 ENOENT (No such file or directory)
14:30:43.876903 openat(AT_FDCWD, "/storage/.kodi/addons/script.module.urllib3/lib/urllib3/packages/rfc3986/abnf_regexp.py", O_RDONLY|O_LARGEFILE) = 10
14:30:43.877117 fstat64(10, {st_mode=S_IFREG|0644, st_size=9081, ...}) = 0
14:30:43.877283 openat(AT_FDCWD, "/storage/.kodi/addons/script.module.urllib3/lib/urllib3/packages/rfc3986/abnf_regexp.pyo", O_RDONLY|O_LARGEFILE) = 12
14:30:43.877455 fstat64(12, {st_mode=S_IFREG|0644, st_size=4905, ...}) = 0
14:30:43.877642 read(12, "\3\363\r\n\217\236\313\\c\0\0\0\0\0\0\0\0\n\0\0\0@\0\0\0s\204\4\0\0d\0"..., 4096) = 4096
14:30:43.877834 fstat64(12, {st_mode=S_IFREG|0644, st_size=4905, ...}) = 0
14:30:43.878629 read(12, "_ABSOLUTEt\f\0\0\0PATH_ABEMPTYt\7\0\0\0P"..., 4096) = 809
14:30:43.878780 read(12, "", 4096)      = 0
14:30:43.879462 close(12)               = 0
14:30:43.880560 close(10)               = 0
14:30:44.097604 mmap2(NULL, 221184, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) = 0x76689000
14:30:44.098639 brk(0x15a6000)          = 0x15a6000
14:30:44.099589 brk(0x15c7000)          = 0x15c7000
14:30:44.100526 brk(0x15e8000)          = 0x15e8000
14:30:44.101381 brk(0x1609000)          = 0x1609000
14:30:44.102300 brk(0x162a000)          = 0x162a000
14:30:44.103167 brk(0x1655000)          = 0x1655000
14:30:44.103618 mmap2(NULL, 221184, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) = 0x76653000
14:30:44.105112 brk(0x1676000)          = 0x1676000
14:30:44.106240 brk(0x1697000)          = 0x1697000
14:30:44.107086 brk(0x16b8000)          = 0x16b8000
14:30:44.107945 brk(0x16d9000)          = 0x16d9000
14:30:44.111069 munmap(0x76689000, 221184) = 0
14:30:44.144624 munmap(0x76653000, 221184) = 0
14:30:44.146260 mmap2(NULL, 266240, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) = 0x7667e000
14:30:44.150173 mmap2(NULL, 266240, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) = 0x7663d000
14:30:44.153618 brk(0x16fa000)          = 0x16fa000
14:30:44.154560 brk(0x171b000)          = 0x171b000
14:30:44.158052 munmap(0x7667e000, 266240) = 0
14:30:44.158637 brk(0x1766000)          = 0x1766000
14:30:44.165449 munmap(0x7663d000, 266240) = 0
14:30:44.166008 brk(0x17a6000)          = 0x17a6000
14:30:44.240580 brk(0x1726000)          = 0x1726000
14:30:44.242240 brk(0x175c000)          = 0x175c000
14:30:44.245815 brk(0x17b2000)          = 0x17b2000
14:30:44.374181 brk(0x1726000)          = 0x1726000
14:30:44.414987 brk(0x175c000)          = 0x175c000
14:30:44.418721 brk(0x17b2000)          = 0x17b2000
14:30:44.546747 brk(0x1726000)          = 0x1726000
14:30:44.572859 brk(0x175c000)          = 0x175c000
14:30:44.576452 brk(0x17b2000)          = 0x17b2000
14:30:44.705972 brk(0x1726000)          = 0x1726000
14:30:44.707396 brk(0x175c000)          = 0x175c000
14:30:44.711048 brk(0x17b2000)          = 0x17b2000
14:30:44.840213 brk(0x1726000)          = 0x1726000
14:30:44.841469 brk(0x175c000)          = 0x175c000
14:30:44.844955 brk(0x17b2000)          = 0x17b2000
14:30:44.973770 brk(0x1726000)          = 0x1726000
14:30:44.975214 brk(0x175c000)          = 0x175c000
14:30:44.978809 brk(0x17b2000)          = 0x17b2000
14:30:45.104792 mmap2(NULL, 262144, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) = 0x7667f000
14:30:45.108075 brk(0x1726000)          = 0x1726000
14:30:45.109371 brk(0x175c000)          = 0x175c000
14:30:45.113320 brk(0x17b2000)          = 0x17b2000
14:30:45.240933 brk(0x1726000)          = 0x1726000
14:30:45.242222 brk(0x175c000)          = 0x175c000
14:30:45.245609 brk(0x17b2000)          = 0x17b2000
14:30:45.371313 brk(0x1726000)          = 0x1726000
14:30:45.372598 brk(0x175c000)          = 0x175c000
14:30:45.376263 brk(0x17b2000)          = 0x17b2000
14:30:45.502044 brk(0x1726000)          = 0x1726000
14:30:45.503598 munmap(0x7667f000, 262144) = 0
14:30:45.506145 brk(0x175c000)          = 0x175c000
14:30:45.509612 brk(0x17b2000)          = 0x17b2000
14:30:45.652234 brk(0x1726000)          = 0x1726000
14:30:45.655678 brk(0x175c000)          = 0x175c000
14:30:45.659229 brk(0x17b2000)          = 0x17b2000
14:30:45.785649 brk(0x1726000)          = 0x1726000
14:30:45.800103 brk(0x175c000)          = 0x175c000
14:30:45.803692 brk(0x17b2000)          = 0x17b2000
14:30:45.929280 brk(0x1726000)          = 0x1726000
14:30:45.930763 brk(0x175c000)          = 0x175c000
14:30:45.934422 brk(0x17b2000)          = 0x17b2000
14:30:46.060708 brk(0x1726000)          = 0x1726000
14:30:46.062034 brk(0x175c000)          = 0x175c000
14:30:46.065457 brk(0x17b2000)          = 0x17b2000
14:30:46.191340 brk(0x1726000)          = 0x1726000
14:30:46.192652 brk(0x175c000)          = 0x175c000
14:30:46.196268 brk(0x17b2000)          = 0x17b2000
14:30:46.322046 mmap2(NULL, 262144, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) = 0x7667f000
14:30:46.325482 brk(0x1726000)          = 0x1726000
14:30:46.326810 brk(0x175c000)          = 0x175c000
14:30:46.331121 brk(0x17b2000)          = 0x17b2000
14:30:46.463667 brk(0x1726000)          = 0x1726000
14:30:46.465295 brk(0x175c000)          = 0x175c000
14:30:46.469072 brk(0x17b2000)          = 0x17b2000
14:30:46.619327 brk(0x1726000)          = 0x1726000
14:30:46.620836 brk(0x175c000)          = 0x175c000
14:30:46.624361 brk(0x17b2000)          = 0x17b2000
14:30:46.757032 brk(0x1726000)          = 0x1726000
14:30:46.758701 munmap(0x7667f000, 262144) = 0
14:30:46.772052 brk(0x175c000)          = 0x175c000
14:30:46.775853 brk(0x17b2000)          = 0x17b2000
14:30:46.908313 brk(0x1726000)          = 0x1726000
14:30:46.909764 brk(0x175c000)          = 0x175c000
14:30:46.913168 brk(0x17b2000)          = 0x17b2000
14:30:47.043481 brk(0x1726000)          = 0x1726000
14:30:47.044705 brk(0x175c000)          = 0x175c000
14:30:47.048061 brk(0x17b2000)          = 0x17b2000
14:30:47.175464 brk(0x1726000)          = 0x1726000
14:30:47.176723 brk(0x175c000)          = 0x175c000
14:30:47.180118 brk(0x17b2000)          = 0x17b2000
14:30:47.304782 mmap2(NULL, 262144, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) = 0x7667f000
14:30:47.308001 brk(0x1726000)          = 0x1726000
14:30:47.309287 brk(0x175c000)          = 0x175c000
14:30:47.312645 brk(0x17b2000)          = 0x17b2000
14:30:47.439730 brk(0x1726000)          = 0x1726000
14:30:47.441109 brk(0x175c000)          = 0x175c000
14:30:47.444652 brk(0x17b2000)          = 0x17b2000
14:30:47.588301 brk(0x1726000)          = 0x1726000
14:30:47.589782 munmap(0x7667f000, 262144) = 0
14:30:47.590138 close(9)                = 0
14:30:47.590410 stat64("/storage/.kodi/addons/script.module.urllib3/lib/urllib3/packages/rfc3986/normalizers", 0x7ebc78e8) = -1 ENOENT (No such file or directory)
14:30:47.590597 openat(AT_FDCWD, "/storage/.kodi/addons/script.module.urllib3/lib/urllib3/packages/rfc3986/normalizers.so", O_RDONLY|O_LARGEFILE) = -1 ENOENT (No such file or directory)
14:30:47.590756 openat(AT_FDCWD, "/storage/.kodi/addons/script.module.urllib3/lib/urllib3/packages/rfc3986/normalizersmodule.so", O_RDONLY|O_LARGEFILE) = -1 ENOENT (No such file or directory)
14:30:47.590900 openat(AT_FDCWD, "/storage/.kodi/addons/script.module.urllib3/lib/urllib3/packages/rfc3986/normalizers.py", O_RDONLY|O_LARGEFILE) = 9
14:30:47.591036 fstat64(9, {st_mode=S_IFREG|0644, st_size=5259, ...}) = 0
14:30:47.591161 openat(AT_FDCWD, "/storage/.kodi/addons/script.module.urllib3/lib/urllib3/packages/rfc3986/normalizers.pyo", O_RDONLY|O_LARGEFILE) = 10
14:30:47.591290 fstat64(10, {st_mode=S_IFREG|0644, st_size=5093, ...}) = 0
14:30:47.591405 read(10, "\3\363\r\n\217\236\313\\c\0\0\0\0\0\0\0\0\2\0\0\0@\0\0\0s\250\0\0\0d\0"..., 4096) = 4096
14:30:47.591529 fstat64(10, {st_mode=S_IFREG|0644, st_size=5093, ...}) = 0
14:30:47.591647 read(10, "\0|\3\0|\6\0|\6\0d\3\0\27!}\7\0t\n\0|\7\0\203\1\0}\10\0|\4"..., 4096) = 997
14:30:47.591802 read(10, "", 4096)      = 0
14:30:47.592124 close(10)               = 0
14:30:47.593242 close(9)                = 0
```

**Comment by dagwieers** [comment]
Unfortunately the Kodi repositories only hosts up to 2 versions of a module, so it is hard to go back to find when this was introduced exactly.

**Comment by Freso** [comment]
This is most likely because the of the `rfc3986` package inclusion in `urllib3`; see https://github.com/Freso/script.module.urllib3/issues/3 – however, this is an upstream issue and should as such be dealt with upstream. I’m not going to do a major rewrite of extremely popular Python libraries specifically for Kodi.

> Given that an addon that imports requests is called for every menu

This sounds like it could be handled more smartly by Kodi itself, by not unloading and then reloading all the Python libraries on every menu change (if this is really what happens). Have you discussed this approach with Kodi developers?

Either, closing this ticket as there’s not really anything actionable that can be done on my end. Steps to decrease load times should happen in Kodi or in [`requests`](https://github.com/kennethreitz/requests) or in whatever of `requests`’s dependencies cause these load times. You may want to give https://github.com/urllib3/urllib3/issues/1590 a read.

**Comment by Freso** [comment]
Looks like you already filed this upstream: https://github.com/kennethreitz/requests/issues/5080

**Comment by dagwieers** [comment]
Yup, but I doubt upstream is going to do anything about this.

BTW I assumed these packages were made specifically for Kodi ? Are they not ?

**Comment by Freso** [comment]
They are, but they are literally just unpacking of official releases moulded to fit Kodi’s add‐on system. There are no code changes compared to upstream.

**Comment by dagwieers** [comment]
> This sounds like it could be handled more smartly by Kodi itself, by not unloading and then reloading all the Python libraries on every menu change (if this is really what happens). Have you discussed this approach with Kodi developers?

There are so many things wrong with the Kodi addon interface, but that's a whole different story.

Fact is that a lot of people are affected by how this works now with the stuff they are running now, and this happened because newer libraries were pushed. Do we have older versions available somewhere so I can pinpoint which version broke this ?

**Comment by Freso** [comment]
Try and switch `script.module.urllib3` to 1.24.1. Like I said before, it is almost certainly not `requests` being the culprit here. (Edit: https://github.com/Freso/script.module.urllib3/archive/2fd0096acdd0d9903baf34daf5c40beb4c8c001d.zip is said version, in case it can’t be retrieved via Kodi’s interface anymore.)

**Comment by dagwieers** [comment]
It is clear what the problem is, things are updated automagically and pushed to hundreds of thousands of devices affecting millions :-) This ought to work better/different IMO.

**Comment by dagwieers** [comment]
> Try and switch script.module.urllib3 to 1.24.1. 

Yes this works.
I couldn't find any other versions yesterday except 1.25.1 and 1.25.2.

Is there a way to revert going to 1.25 ?
Is there a specific reason we update to the newer/latest version ?

**Comment by Freso** [comment]
> Is there a way to revert going to 1.25 ?

I didn’t package 1.25.0 for Kodi, because 1) by the time I got to it, 1.25.1 was released, 2) 1.25.0 had some bad bugs so that even [upstream `urllib3` recommends `requests` to specifically not support 1.25.0](https://github.com/kennethreitz/requests/pull/5063#issuecomment-486346386) when they change to officially support urllib3 1.25.x.

Additionally, 1.25.0 also includes the `rfc3986` that is causing this, so it would still have the load time issue you’re facing.

>  Is there a specific reason we update to the newer/latest version ?

`urllib3` 1.25.0 fixed a [security flaw](https://nvd.nist.gov/vuln/detail/CVE-2019-11324) that exists in `urllib3` ≤ 1.24.2.

Anyway, we’re now discussing the `script.module.urllib3` Kodi add‐on and upstream `urllib3` issues and decisions, not things for `script.module.requests` per se. This issue is not an issue for `script.module.requests`. It is an issue for upstream `urllib3` and *possibly* for `script.module.urllib3` and/or upstream `requests`. All 3 of which have existing relevant tickets. Please take any further questions/discussion there.
